File: app/src/main/java/org/dhis2/usescases/teiDashboard/adapters/DashboardPagerAdapter.java
package org.dhis2.usescases.teiDashboard.adapters;

import android.content.Context;
import android.os.Parcelable;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;

import org.dhis2.R;
import org.dhis2.usescases.teiDashboard.dashboardfragments.indicators.IndicatorsFragment;
import org.dhis2.usescases.notes.NotesFragment;
import org.dhis2.usescases.teiDashboard.dashboardfragments.relationships.RelationshipFragment;
import org.dhis2.usescases.teiDashboard.dashboardfragments.tei_data.TEIDataFragment;
import org.jetbrains.annotations.NotNull;

/**
 * QUADRAM. Created by ppajuelo on 29/11/2017.
 */

public class DashboardPagerAdapter extends FragmentStatePagerAdapter {

    private static final int MOBILE_DASHBOARD_SIZE = 4;
    private final String enrollmentUid;
    private String currentProgram;
    private final String teiUid;
    private Context context;


    public DashboardPagerAdapter(Context context, FragmentManager fm, String program, String teiUid, String enrollmentUid) {
        super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        this.context = context;
        this.currentProgram = program;
        this.teiUid = teiUid;
        this.enrollmentUid = enrollmentUid;
    }

    @Override
    public Parcelable saveState() {
        // Do Nothing
        return null;
    }

    private IndicatorsFragment indicatorsFragment;
    private RelationshipFragment relationshipFragment;
    private TEIDataFragment teiDataFragment;

    @NotNull
    @Override
    public Fragment getItem(int position) {
        switch (position) {
            case 1:
                if (indicatorsFragment == null) {
                    indicatorsFragment = new IndicatorsFragment();
                }
                return indicatorsFragment;
            case 2:
                if (relationshipFragment == null) {
                    relationshipFragment = new RelationshipFragment();
                }
                return relationshipFragment;
            case 3:
                return NotesFragment.newTrackerInstance(currentProgram, teiUid);
            default:
                return TEIDataFragment.newInstance(currentProgram, teiUid, enrollmentUid);
        }
    }

    @Override
    public int getCount() {
        return currentProgram != null ? MOBILE_DASHBOARD_SIZE : 1;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        switch (position) {
            default:
                return context.getString(R.string.dashboard_overview);
            case 1:
                return context.getString(R.string.dashboard_indicators);
            case 2:
                return context.getString(R.string.dashboard_relationships);
            case 3:
                return context.getString(R.string.dashboard_notes);
        }
    }
}

File: app/src/main/java/org/dhis2/usescases/teiDashboard/adapters/DashboardPagerAdapter.kt
package org.dhis2.usescases.teiDashboard.adapters

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import org.dhis2.R
import org.dhis2.usescases.notes.NotesFragment
import org.dhis2.usescases.teiDashboard.dashboardfragments.indicators.IndicatorsFragment
import org.dhis2.usescases.teiDashboard.dashboardfragments.indicators.VISUALIZATION_TYPE
import org.dhis2.usescases.teiDashboard.dashboardfragments.indicators.VisualizationType
import org.dhis2.usescases.teiDashboard.dashboardfragments.relationships.RelationshipFragment
import org.dhis2.usescases.teiDashboard.dashboardfragments.teidata.TEIDataFragment
import org.dhis2.utils.isLandscape

class DashboardPagerAdapter(
    private val fragmentActivity: FragmentActivity,
    private val currentProgram: String?,
    private val teiUid: String,
    private val enrollmentUid: String?,
    private val displayAnalyticScreen: Boolean = true,
    private val displayRelationshipScreen: Boolean,
) : FragmentStateAdapter(fragmentActivity) {

    enum class DashboardPageType {
        TEI_DETAIL, ANALYTICS, RELATIONSHIPS, NOTES
    }

    private var indicatorsFragment: IndicatorsFragment? = null
    private var relationshipFragment: RelationshipFragment? = null
    private val landscapePages: List<DashboardPageType>
    private val portraitPages: List<DashboardPageType>

    init {
        landscapePages = mutableListOf<DashboardPageType>().apply {
            if (displayAnalyticScreen) add(DashboardPageType.ANALYTICS)
            if (displayRelationshipScreen) add(DashboardPageType.RELATIONSHIPS)
            if (currentProgram != null) add(DashboardPageType.NOTES)
        }
        portraitPages = mutableListOf<DashboardPageType>().apply {
            add(DashboardPageType.TEI_DETAIL)
            if (displayAnalyticScreen) add(DashboardPageType.ANALYTICS)
            if (displayRelationshipScreen) add(DashboardPageType.RELATIONSHIPS)
            if (currentProgram != null) add(DashboardPageType.NOTES)
        }
    }

    override fun createFragment(position: Int): Fragment {
        return createFragmentForPage(
            if (fragmentActivity.isLandscape()) {
                landscapePages[position]
            } else {
                portraitPages[position]
            },
        )
    }

    private fun createFragmentForPage(pageType: DashboardPageType): Fragment {
        return when (pageType) {
            DashboardPageType.TEI_DETAIL -> TEIDataFragment.newInstance(
                currentProgram,
                teiUid,
                enrollmentUid,
            )
            DashboardPageType.ANALYTICS -> {
                if (indicatorsFragment == null) {
                    indicatorsFragment = IndicatorsFragment().apply {
                        arguments = Bundle().apply {
                            putString(VISUALIZATION_TYPE, VisualizationType.TRACKER.name)
                        }
                    }
                }
                indicatorsFragment!!
            }
            DashboardPageType.RELATIONSHIPS -> {
                if (relationshipFragment == null) {
                    relationshipFragment = RelationshipFragment().apply {
                        arguments = RelationshipFragment.withArguments(
                            currentProgram,
                            teiUid,
                            enrollmentUid,
                            null,
                        )
                    }
                }
                relationshipFragment!!
            }
            DashboardPageType.NOTES -> NotesFragment.newTrackerInstance(currentProgram!!, teiUid)
        }
    }

    override fun getItemCount() =
        if (fragmentActivity.isLandscape()) landscapePages.size else portraitPages.size

    fun getNavigationPagePosition(navigationId: Int): Int {
        val pageType = when (navigationId) {
            R.id.navigation_details -> DashboardPageType.TEI_DETAIL
            R.id.navigation_analytics -> DashboardPageType.ANALYTICS
            R.id.navigation_relationships -> DashboardPageType.RELATIONSHIPS
            R.id.navigation_notes -> DashboardPageType.NOTES
            R.id.navigation_events -> null
            else -> null
        }

        return pageType?.let {
            if (fragmentActivity.isLandscape()) {
                landscapePages.indexOf(pageType)
            } else {
                portraitPages.indexOf(pageType)
            }
        } ?: NO_POSITION
    }

    fun pageType(position: Int): DashboardPageType {
        return if (fragmentActivity.isLandscape()) {
            landscapePages[position]
        } else {
            portraitPages[position]
        }
    }

    companion object {
        const val NO_POSITION = -1
    }
}
